README.rst
========
ciso8601
========

.. image:: https://circleci.com/gh/closeio/ciso8601/tree/master.svg?style=svg&circle-token=72fc522063916cb1c6c5c9882b97db9d2ed651d8
    :target: https://circleci.com/gh/closeio/ciso8601/tree/master

``ciso8601`` converts `ISO 8601`_ date time strings into Python datetime objects.
Since it's written as a C module, it is much faster than other Python libraries.
Tested with Python 2.7, 3.4, 3.5, 3.6, 3.7b.

.. _ISO 8601: https://en.wikipedia.org/wiki/ISO_8601

(Interested in working on projects like this? `Close.io`_ is looking for `great engineers`_ to join our team)

.. _Close.io: https://close.io
.. _great engineers: https://jobs.close.io


.. contents:: Contents


Quick Start
-----------

.. code:: bash

  % pip install ciso8601

.. code:: python

  In [1]: import ciso8601

  In [2]: ciso8601.parse_datetime('2014-12-05T12:30:45.123456-05:30')
  Out[2]: datetime.datetime(2014, 12, 5, 12, 30, 45, 123456, tzinfo=pytz.FixedOffset(330))

  In [3]: ciso8601.parse_datetime('20141205T123045')
  Out[3]: datetime.datetime(2014, 12, 5, 12, 30, 45)

Migration to v2
---------------

Version 2.0.0 of ``ciso8601`` changed the core implementation. This was not entirely backwards compatible, and care should be taken when migrating
See `CHANGELOG`_ for the Migration Guide.

.. _CHANGELOG: https://github.com/closeio/ciso8601/blob/master/CHANGELOG.md

Error Handling
--------------

Starting in v2.0.0, ``ciso8601`` offers strong guarantees when it comes to parsing strings.

``parse_datetime(dt: String): datetime`` is a function that takes a string and either:

* Returns a properly parsed Python datetime, **if and only if** the **entire** string conforms to the supported subset of ISO 8601
* Raises a ``ValueError`` with a description of the reason why the string doesn't conform to the supported subset of ISO 8601

If time zone information is provided, an aware datetime object will be returned. Otherwise, a naive datetime is returned.

Benchmark
---------

Date time string with no time zone information:

.. code:: python

  In [1]: import datetime, aniso8601, iso8601, isodate, dateutil.parser, arrow, ciso8601

  In [2]: ds = u'2014-01-09T21:48:00.921000'

  In [3]: %timeit ciso8601.parse_datetime(ds)
  1000000 loops, best of 3: 204 ns per loop

  In [4]: %timeit datetime.datetime.strptime(ds, "%Y-%m-%dT%H:%M:%S.%f")
  100000 loops, best of 3: 15 µs per loop

  In [5]: %timeit dateutil.parser.parse(ds)
  10000 loops, best of 3: 122 µs per loop

  In [6]: %timeit aniso8601.parse_datetime(ds)
  10000 loops, best of 3: 28.9 µs per loop

  In [7]: %timeit iso8601.parse_date(ds)
  10000 loops, best of 3: 42 µs per loop

  In [8]: %timeit isodate.parse_datetime(ds)
  10000 loops, best of 3: 69.4 µs per loop

  In [9]: %timeit arrow.get(ds).datetime
  10000 loops, best of 3: 87 µs per loop

ciso8601 takes 0.204us, which is 73x faster than datetime's strptime, which is
not a full ISO8601 parser. It is **141x faster than aniso8601**, the next fastest
ISO8601 parser in this comparison.

Date time string with time zone information:

.. code:: python

  In [1]: import datetime, aniso8601, iso8601, isodate, dateutil.parser, arrow, ciso8601

  In [2]: ds = u'2014-01-09T21:48:00.921000+05:30'

  In [3]: %timeit ciso8601.parse_datetime(ds)
  1000000 loops, best of 3: 525 ns per loop

  In [4]: %timeit dateutil.parser.parse(ds)
  10000 loops, best of 3: 162 µs per loop

  In [5]: %timeit aniso8601.parse_datetime(ds)
  10000 loops, best of 3: 36.8 µs per loop

  In [6]: %timeit iso8601.parse_date(ds)
  10000 loops, best of 3: 53.5 µs per loop

  In [7]: %timeit isodate.parse_datetime(ds)
  10000 loops, best of 3: 82.6 µs per loop

  In [8]: %timeit arrow.get(ds).datetime
  10000 loops, best of 3: 104 µs per loop

Even with time zone information, ``ciso8601`` is 70x as fast as ``aniso8601``.

Tested on Python 2.7.10 on macOS 10.12.6 using the following modules:

.. code:: python

  aniso8601==1.2.1
  arrow==0.10.0
  ciso8601==1.0.4
  iso8601==0.1.12
  isodate==0.5.4
  python-dateutil==2.6.1

Supported Subset of ISO 8601
----------------------------

``ciso8601`` only supports the most common subset of ISO 8601.

Date Formats
^^^^^^^^^^^^

The following date formats are supported:

.. table::
   :widths: auto

   ============================= ============== ==================
   Format                        Example        Supported
   ============================= ============== ==================
   ``YYYY-MM-DD``                ``2018-04-29`` ✅
   ``YYYY-MM``                   ``2018-04``    ✅
   ``YYYYMMDD``                  ``2018-04``    ✅
   ``--MM-DD`` (omitted year)    ``--04-29``    ❌              
   ``--MMDD`` (omitted year)     ``--0429``     ❌
   ``±YYYYY-MM`` (>4 digit year) ``+10000-04``  ❌   
   ``+YYYY-MM`` (leading +)      ``+2018-04``   ❌   
   ``-YYYY-MM`` (negative -)     ``-2018-04``   ❌   
   ============================= ============== ==================

Week dates or ordinal dates are not currently supported.

.. table::
   :widths: auto

   ============================= ============== ==================
   Format                        Example        Supported
   ============================= ============== ==================
   ``YYYY-Www`` (week date)      ``2009-W01``   ❌
   ``YYYYWww`` (week date)       ``2009W01``    ❌
   ``YYYY-Www-D`` (week date)    ``2009-W01-1`` ❌
   ``YYYYWwwD`` (week date)      ``2009-W01-1`` ❌
   ``YYYY-DDD`` (ordinal date)   ``1981-095``   ❌
   ``YYYYDDD`` (ordinal date)    ``1981095``    ❌ 
   ============================= ============== ==================

Time Formats
^^^^^^^^^^^^

Times are optional and are separated from the date by the letter ``T``.

Consistent with `RFC 3339`_, ``ciso860`` also allows either a space character, or a lower-case ``t``, to be used instead of a ``T``.

.. _RFC 3339: https://stackoverflow.com/questions/522251/whats-the-difference-between-iso-8601-and-rfc-3339-date-formats)

The following time formats are supported:

.. table::
   :widths: auto

   =================================== =================== ==============  
   Format                              Example             Supported          
   =================================== =================== ============== 
   ``hh``                              ``11``              ✅ 
   ``hhmm``                            ``1130``            ✅ 
   ``hh:mm``                           ``11:30``           ✅ 
   ``hhmmss``                          ``113059``          ✅ 
   ``hh:mm:ss``                        ``11:30:59``        ✅ 
   ``hhmmss.ssssss``                   ``113059.123456``   ✅ 
   ``hh:mm:ss.ssssss``                 ``11:30:59.123456`` ✅ 
   ``hhmmss,ssssss``                   ``113059,123456``   ✅ 
   ``hh:mm:ss,ssssss``                 ``11:30:59,123456`` ✅ 
   Midnight (special case)             ``24:00:00``        ✅               
   ``hh.hhh`` (fractional hours)       ``11.5``            ❌               
   ``hh:mm.mmm`` (fractional minutes)  ``11:30.5``         ❌               
   =================================== =================== ============== 

**Note:** Python datetime objects only have microsecond precision (6 digits). Any additional precision will be truncated.

Time Zone Information
^^^^^^^^^^^^^^^^^^^^^

Time zone information may be provided in one of the following formats:

.. table::
   :widths: auto

   ========== ========== =========== 
   Format     Example    Supported          
   ========== ========== =========== 
   ``Z``      ``Z``      ✅
   ``z``      ``z``      ✅
   ``±hh``    ``+11``    ✅
   ``±hhmm``  ``+1130``  ✅
   ``±hh:mm`` ``+11:30`` ✅
   ========== ========== ===========

While the ISO 8601 specification allows the use of MINUS SIGN (U+2212) in the time zone separator, ``ciso8601`` only supports the use of the HYPHEN-MINUS (U+002D) character.

Consistent with `RFC 3339`_, ``ciso860`` also allows a lower-case ``z`` to be used instead of a ``Z``.

Ignoring Timezone Information While Parsing
-------------------------------------------

It takes more time to parse timestamps with time zone information, especially if they're not in UTC. However, there are times when you don't care about time zone information, and wish to produce naive datetimes instead.
For example, if you are certain that your program will only parse timestamps from a single time zone, you might want to strip the time zone information and only output naive datetimes.

In these limited cases, there is a second function provided.
``parse_datetime_as_naive`` will ignore any time zone information it finds and, as a result, is faster for timestamps containing time zone information.

.. code:: python

  In [1]: import ciso8601

  In [2]: ciso8601.parse_datetime_as_naive('2014-12-05T12:30:45.123456-05:30')
  Out[2]: datetime.datetime(2014, 12, 5, 12, 30, 45, 123456)

NOTE: ``parse_datetime_as_naive`` is only useful in the case where your timestamps have time zone information, but you want to ignore it. This is somewhat unusual.
If your timestamps don't have time zone information (i.e. are naive), simply use ``parse_datetime``. It is just as fast.


CHANGELOG.md
<!-- Generated with "Markdown T​O​C" extension for Visual Studio Code -->
<!-- TOC anchorMode:github.com -->

- [2.x.x](#2xx)
    - [Version 2.0.1](#version-201)
    - [Version 2.0.0](#version-200)
        - [Breaking changes](#breaking-changes)
        - [Other Changes](#other-changes)
        - [v1.x.x -> 2.0.0 Migration guide](#v1xx---200-migration-guide)
            - [ValueError instead of None](#valueerror-instead-of-none)
            - [Tightened ISO 8601 conformance](#tightened-iso-8601-conformance)
            - [`parse_datetime_unaware` has been renamed](#parse_datetime_unaware-has-been-renamed)

<!-- /TOC -->

# 2.x.x

## Version 2.1.0

* Switched to using a C implementation of `timezone` objects.
    * Much faster parse times for timestamps with timezone information
        * ~3x faster on Python 2.7, 25% faster on Python 3.7
    * Thanks to [`pendulum`](https://github.com/sdispater/pendulum) and @sdispater for the code.
    * Python 2.7 users no longer need to install `pytz` dependency :smiley:

## Version 2.0.1

* Fixed some memory leaks introduced in 2.0.0 (#51)

## Version 2.0.0

Version 2.0.0 was a major rewrite of `ciso8601`.

Version 1.x.x had a problem with error handling in the case of invalid timestamps.

In 1.x.x, parse_datetime:

* All valid datetime strings within the supported subset of ISO 8601 would result in the correct Python datetime (this was good)
* Some invalid timestamps will return `None` and others might get truncated and return an incorrect Python datetime (this was bad)

A developer with a given timestamp string, could not predict a priori what `ciso8601` is going to return without looking at the code.
Fundamentally, this is the problem that version 2 addressed.

Fundamentally, `parse_datetime(dt: String): datetime` was rewritten so that it takes a string and either:

* Returns a properly parsed Python datetime, **if and only if** that **entire** string conforms to the supported subset of ISO 8601
* Raises an `ValueError` with a description of the reason why the string doesn't conform to the supported subset of ISO 8601

### Breaking changes

1. Version 2 now raises `ValueError` when a timestamp does not conform to the supported subset of ISO 8601
    * This includes trailing characters in the timestamp
    * No longer accepts single character "day" values
    * See migration guide below for more examples
1. `parse_datetime_unaware` was renamed to `parse_datetime_as_naive` (See "Migration Guide" below for reasons)

### Other Changes

* Attempting to parse a timestamp with time zone information without having pytz installed raises `ImportError` (Only affects Python 2.7). Fixes #19
* Added support for the special case of midnight (24:00:00) that is valid in ISO 8601. Fixes #41
* Fixed bug where "20140200" would not fail, but produce 2014-02-01. Fixes #42

### v1.x.x -> 2.0.0 Migration guide

#### ValueError instead of None

Places where you were checking for a return of `None` from ciso8601:

```python
    timestamp = "2018-01-01T00:00:00+05:00"
    dt = parse_datetime(timestamp)
    if dt is None:
        raise ValueError(f"Could not parse {timestamp}")
```

You should change to now expect `ValueError` to be thrown:

```python
    timestamp = "2018-01-01T00:00:00+05:00"
    dt = parse_datetime(timestamp)
```

#### Tightened ISO 8601 conformance

The rules with respect to what ciso8601 will consider a conforming ISO 8601 string have been tightened.

Now a timestamp will parse **if and only if** the timestamp is 100% conforming to the supported subset of the ISO 8601 specification.


```python
    # trailing separator
    "2014-"
    "2014-01-"
    "2014-01-01T"
    "2014-01-01T00:"
    "2014-01-01T00:00:"
    "2014-01-01T00:00:00-"
    "2014-01-01T00:00:00-00:"

    # Mix of no-separator and separator
    "201401-02" 
    "2014-0102"
    "2014-01-02T00:0000" 
    "2014-01-02T0000:00"
    
    "2014-01-02T01:23:45Zabcdefghij" # Trailing characters

    "2014-01-1" # Single digit day
    "2014-01-01T00:00:00-0:04" # Single digit tzhour
    "2014-01-01T00:00:00-00:4" # Single digit tzminute
```

These should have been considered bugs in ciso8601 1.x.x, but it may be the case that your code was relying on the previously lax parsing rules.

#### `parse_datetime_unaware` has been renamed

`parse_datetime_unaware` existed for the case where your input timestamp had time zone information, but you wanted to ignore the time zone information and therefore could save some cycles by not creating the underlying `tzinfo` object.

It has been renamed to `parse_datetime_as_naive` for 2 reasons:

1. Developers were assuming that `parse_datetime_unaware` was the function to use for parsing naive timestamps, when really it is for parsing timestamps with time zone information as naive datetimes. `parse_datetime` handles parsing both timestamps with and without time zone information and should be used for all parsing, unless you actually need this use case. See additional description in [the README](https://github.com/closeio/ciso8601/tree/raise-valueerror-on-invalid-dates#ignoring-timezone-information-while-parsing) for a more detailed description of this use case.
2. Python [refers to datetimes without time zone information](https://docs.python.org/3/library/datetime.html) as `naive`, not `unaware` 

Before switching all instances of `parse_datetime_unaware`, make sure to ask yourself whether you actually intended to use `parse_datetime_unaware`.

* If you meant to parse naive timestamps as naive datetimes, use `parse_datetime` instead.
* If you actually meant to parse **timestamps with time zone information** as **naive** datetimes, use `parse_datetime_as_naive` instead.

|                                    | Input with TZ Info | Input without TZ Info |
| ---------------------------------- | ------------------ | --------------------- |
| `parse_datetime()` output          | tz aware datetime  | tz naive datetime     |

module.c
#include <Python.h>
#include <datetime.h>

#include "timezone.h"

#define PY_VERSION_AT_LEAST_36 \
    ((PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION >= 6) || PY_MAJOR_VERSION > 3)

static PyObject *utc;

#define PARSE_INTEGER(field, length, field_name)                       \
    for (i = 0; i < length; i++) {                                     \
        if (*c >= '0' && *c <= '9') {                                  \
            field = 10 * field + *c++ - '0';                           \
        }                                                              \
        else {                                                         \
            return format_unexpected_character_exception(              \
                field_name, *c, (c - str) / sizeof(char), length - i); \
        }                                                              \
    }

#define PARSE_FRACTIONAL_SECOND()                              \
    for (i = 0; i < 6; i++) {                                  \
        if (*c >= '0' && *c <= '9') {                          \
            usecond = 10 * usecond + *c++ - '0';               \
        }                                                      \
        else if (i == 0) {                                     \
            /* We need at least one digit. */                  \
            /* Trailing '.' or ',' is not allowed */           \
            return format_unexpected_character_exception(      \
                "subsecond", *c, (c - str) / sizeof(char), 1); \
        }                                                      \
        else                                                   \
            break;                                             \
    }                                                          \
                                                               \
    /* Omit excessive digits */                                \
    while (*c >= '0' && *c <= '9') c++;                        \
                                                               \
    /* If we break early, fully expand the usecond */          \
    while (i++ < 6) usecond *= 10;

#define PARSE_SEPARATOR(separator, field_name)                                \
    if (separator) {                                                          \
        c++;                                                                  \
    }                                                                         \
    else {                                                                    \
        PyErr_Format(PyExc_ValueError,                                        \
                     "Invalid character while parsing %s ('%c', Index: %lu)", \
                     field_name, *c, (c - str) / sizeof(char));               \
        return NULL;                                                          \
    }

static void *
format_unexpected_character_exception(char *field_name, char c, size_t index,
                                      int expected_character_count)
{
    if (c == '\0')
        PyErr_Format(
            PyExc_ValueError,
            "Unexpected end of string while parsing %s. Expected %d more "
            "character%s",
            field_name, expected_character_count,
            (expected_character_count != 1) ? "s" : "");
    else
        PyErr_Format(PyExc_ValueError,
                     "Invalid character while parsing %s ('%c', Index: %zu)",
                     field_name, c, index);
    return NULL;
}

#define IS_CALENDAR_DATE_SEPARATOR (*c == '-')
#define IS_DATE_AND_TIME_SEPARATOR (*c == 'T' || *c == ' ' || *c == 't')
#define IS_TIME_SEPARATOR (*c == ':')
#define IS_TIME_ZONE_SEPARATOR \
    (*c == 'Z' || *c == '-' || *c == '+' || *c == 'z')
#define IS_FRACTIONAL_SEPARATOR (*c == '.' || *c == ',')

static PyObject *
_parse(PyObject *self, PyObject *args, int parse_any_tzinfo)
{
    PyObject *obj;
    PyObject *tzinfo = Py_None;

    int i;
    char *str = NULL;
    char *c;
    int year = 0, month = 0, day = 0, hour = 0, minute = 0, second = 0,
        usecond = 0;
    int time_is_midnight = 0;
    int tzhour = 0, tzminute = 0, tzsign = 0;
    PyObject *delta;
    PyObject *temp;

    if (!PyArg_ParseTuple(args, "s", &str))
        return NULL;
    c = str;

    /* Year */
    PARSE_INTEGER(year, 4, "year")

#if !PY_VERSION_AT_LEAST_36
    /* Python 3.6+ does this validation as part of Datetime's C API
     * constructor. See
     * https://github.com/python/cpython/commit/b67f0967386a9c9041166d2bbe0a421bd81e10bc
     * We skip ` || year < datetime.MAXYEAR)`, since cio8601 currently doesn't
     * support 5 character years, so it is impossible.
     */
    if (year <
        1) /* datetime.MINYEAR = 1, which is not exposed to the C API. */
    {
        PyErr_Format(PyExc_ValueError, "year %d is out of range", year);
        return NULL;
    }
#endif

    if (IS_CALENDAR_DATE_SEPARATOR) { /* Separated Month and Day (ie. MM-DD) */
        c++;
        /* Month */
        PARSE_INTEGER(month, 2, "month")

        if (*c != '\0' && !IS_DATE_AND_TIME_SEPARATOR) { /* Optional Day */
            PARSE_SEPARATOR(IS_CALENDAR_DATE_SEPARATOR, "date separator ('-')")

            /* Day */
            PARSE_INTEGER(day, 2, "day")
        }
        else {
            day = 1;
        }
    }
    else { /* Non-separated Month and Day (ie. MMDD) */
        /* Month */
        PARSE_INTEGER(month, 2, "month")
        /* Note that YYMM is not a valid timestamp. If the calendar date is not
         * separated, a day is required (ie. YYMMDD)
         */
        PARSE_INTEGER(day, 2, "day")
    }

#if !PY_VERSION_AT_LEAST_36
    /* Validation of date fields
     * These checks are needed for Python <3.6 support. See
     * https://github.com/closeio/ciso8601/pull/30 Python 3.6+ does this
     * validation as part of Datetime's C API constructor. See
     * https://github.com/python/cpython/commit/b67f0967386a9c9041166d2bbe0a421bd81e10bc
     */
    if (month < 1 || month > 12) {
        PyErr_SetString(PyExc_ValueError, "month must be in 1..12");
        return NULL;
    }

    if (day < 1) {
        PyErr_SetString(PyExc_ValueError, "day is out of range for month");
        return NULL;
    }

    /* Validate max day based on month */
    switch (month) {
        case 2:
            /* In the Gregorian calendar three criteria must be taken into
             * account to identify leap years:
             *     -The year can be evenly divided by 4;
             *     -If the year can be evenly divided by 100, it is NOT a leap
             * year, unless;
             *     -The year is also evenly divisible by 400. Then it is a leap
             * year.
             */
            if (day > 28) {
                unsigned int leap =
                    (year % 4 == 0) && (year % 100 || (year % 400 == 0));
                if (leap == 0 || day > 29) {
                    PyErr_SetString(PyExc_ValueError,
                                    "day is out of range for month");
                    return NULL;
                }
            }
            break;
        case 4:
        case 6:
        case 9:
        case 11:
            if (day > 30) {
                PyErr_SetString(PyExc_ValueError,
                                "day is out of range for month");
                return NULL;
            }
            break;
        default:
            /* For other months i.e. 1, 3, 5, 7, 8, 10 and 12 */
            if (day > 31) {
                PyErr_SetString(PyExc_ValueError,
                                "day is out of range for month");
                return NULL;
            }
            break;
    }
#endif

    if (*c != '\0') {
        /* Date and time separator */
        PARSE_SEPARATOR(IS_DATE_AND_TIME_SEPARATOR,
                        "date and time separator (ie. 'T' or ' ')")

        /* Hour */
        PARSE_INTEGER(hour, 2, "hour")

        if (*c != '\0' &&
            !IS_TIME_ZONE_SEPARATOR) { /* Optional minute and second */

            if (IS_TIME_SEPARATOR) { /* Separated Minute and Second (ie. mm:ss)
                                      */
                c++;

                /* Minute */
                PARSE_INTEGER(minute, 2, "minute")

                if (*c != '\0' &&
                    !IS_TIME_ZONE_SEPARATOR) { /* Optional Second */
                    PARSE_SEPARATOR(IS_TIME_SEPARATOR, "time separator (':')")

                    /* Second */
                    PARSE_INTEGER(second, 2, "second")

                    /* Optional Fractional Second */
                    if (IS_FRACTIONAL_SEPARATOR) {
                        c++;
                        PARSE_FRACTIONAL_SECOND()
                    }
                }
            }
            else { /* Non-separated Minute and Second (ie. mmss) */
                /* Minute */
                PARSE_INTEGER(minute, 2, "minute")
                if (*c != '\0' &&
                    !IS_TIME_ZONE_SEPARATOR) { /* Optional Second */
                    /* Second */
                    PARSE_INTEGER(second, 2, "second")

                    /* Optional Fractional Second */
                    if (IS_FRACTIONAL_SEPARATOR) {
                        c++;
                        PARSE_FRACTIONAL_SECOND()
                    }
                }
            }
        }

        if (hour == 24 && minute == 0 && second == 0 && usecond == 0) {
            /* Special case of 24:00:00, that is allowed in ISO 8601. It is
             * equivalent to 00:00:00 the following day
             */
            hour = 0, minute = 0, second = 0, usecond = 0;
            time_is_midnight = 1;
        }

#if !PY_VERSION_AT_LEAST_36
        /* Validate hour/minute/second
         * Only needed for Python <3.6 support.
         * Python 3.6+ does this validation as part of Datetime's constructor).
         */
        if (hour > 23) {
            PyErr_SetString(PyExc_ValueError, "hour must be in 0..23");
            return NULL;
        }
        if (minute > 59) {
            PyErr_SetString(PyExc_ValueError, "minute must be in 0..59");
            return NULL;
        }
        if (second > 59) {
            PyErr_SetString(PyExc_ValueError, "second must be in 0..59");
            return NULL;
        }
#endif

        /* Optional tzinfo */
        if (IS_TIME_ZONE_SEPARATOR) {
            if (*c == '+') {
                tzsign = 1;
            }
            else if (*c == '-') {
                tzsign = -1;
            }
            c++;

            if (tzsign != 0) {
                /* tz hour */
                PARSE_INTEGER(tzhour, 2, "tz hour")

                if (IS_TIME_SEPARATOR) { /* Optional separator */
                    c++;
                    /* tz minute */
                    PARSE_INTEGER(tzminute, 2, "tz minute")
                }
                else if (*c != '\0') { /* Optional tz minute */
                    PARSE_INTEGER(tzminute, 2, "tz minute")
                }
            }

            /* It's not entirely clear whether this validation check is
             * necessary under ISO 8601. For now, we will error on the side of
             * caution and prevent suspected invalid timestamps If we need to
             * loosen this restriction later, we can.
             */
            if (tzminute > 59) {
                PyErr_SetString(PyExc_ValueError, "tzminute must be in 0..59");
                return NULL;
            }

            if (parse_any_tzinfo) {
                tzminute += 60 * tzhour;
                tzminute *= tzsign;

                if (tzminute == 0) {
                    tzinfo = utc;
                }
                else if (abs(tzminute) >= 1440) {
                    /* Format the error message as if we were still using pytz
                     * for Python 2 and datetime.timezone for Python 3.
                     * This is done to maintain complete backwards
                     * compatibility with ciso8601 2.0.x. Perhaps change to a
                     * simpler message in ciso8601 v3.0.0.
                     */
#if PY_MAJOR_VERSION >= 3
                    delta = PyDelta_FromDSU(0, tzminute * 60, 0);
                    PyErr_Format(PyExc_ValueError,
                                 "offset must be a timedelta"
                                 " strictly between -timedelta(hours=24) and"
                                 " timedelta(hours=24),"
                                 " not %R.",
                                 delta);
                    Py_DECREF(delta);
#else
                    PyErr_Format(PyExc_ValueError,
                                 "('absolute offset is too large', %d)",
                                 tzminute);
#endif
                    return NULL;
                }
                else {
                    tzinfo = new_fixed_offset(60 * tzminute);
                    if (tzinfo == NULL) /* ie. PyErr_Occurred() */
                        return NULL;
                }
            }
        }
    }

    /* Make sure that there is no more to parse. */
    if (*c != '\0') {
        PyErr_Format(PyExc_ValueError, "unconverted data remains: '%s'", c);
        return NULL;
    }

    obj = PyDateTimeAPI->DateTime_FromDateAndTime(
        year, month, day, hour, minute, second, usecond, tzinfo,
        PyDateTimeAPI->DateTimeType);

    if (tzinfo != Py_None && tzinfo != utc)
        Py_DECREF(tzinfo);

    if (obj && time_is_midnight) {
        delta = PyDelta_FromDSU(1, 0, 0); /* 1 day */
        temp = obj;
        obj = PyNumber_Add(temp, delta);
        Py_DECREF(delta);
        Py_DECREF(temp);
    }

    return obj;
}

static PyObject *
parse_datetime_as_naive(PyObject *self, PyObject *args)
{
    return _parse(self, args, 0);
}

static PyObject *
parse_datetime(PyObject *self, PyObject *args)
{
    return _parse(self, args, 1);
}

static PyMethodDef CISO8601Methods[] = {
    {"parse_datetime", parse_datetime, METH_VARARGS,
     "Parse a ISO8601 date time string."},
    {"parse_datetime_as_naive", parse_datetime_as_naive, METH_VARARGS,
     "Parse a ISO8601 date time string, ignoring the time zone component."},
    {NULL, NULL, 0, NULL}};

#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef moduledef = {
    PyModuleDef_HEAD_INIT,
    "ciso8601",
    NULL,
    -1,
    CISO8601Methods,
    NULL,
    NULL,
    NULL,
    NULL,
};
#endif

PyMODINIT_FUNC
#if PY_MAJOR_VERSION >= 3
PyInit_ciso8601(void)
#else
initciso8601(void)
#endif
{
#if PY_MAJOR_VERSION >= 3
    PyObject *module = PyModule_Create(&moduledef);
#else
    (void)Py_InitModule("ciso8601", CISO8601Methods);
#endif
    PyDateTime_IMPORT;

#if PY_MAJOR_VERSION >= 3
    if (initialize_timezone_code() < 0)
        return NULL;
#else
    initialize_timezone_code();
#endif

    utc = new_fixed_offset(0);

#if PY_MAJOR_VERSION >= 3
    return module;
#endif
}
